modules/apps/dynamic-data-mapping/dynamic-data-mapping-test/src/testIntegration/java/com/liferay/dynamic/data/mapping/upgrade/v5_1_1/test/DDMValidationUpgradeProcessDDMStructureTest.java

```diff
@@ -0,0 +1,157 @@
+import com.liferay.dynamic.data.mapping.test.util.DDMStructureTestUtil;
+import com.liferay.portal.kernel.model.Group;
+import com.liferay.portal.kernel.model.change.tracking.CTModel;
+import com.liferay.portal.kernel.service.change.tracking.CTService;
+import com.liferay.portal.kernel.test.rule.AggregateTestRule;
+import com.liferay.portal.kernel.test.rule.DeleteAfterTestRun;
+import com.liferay.portal.kernel.test.util.GroupTestUtil;
+import com.liferay.portal.kernel.test.util.RandomTestUtil;
+import com.liferay.portal.kernel.upgrade.UpgradeProcess;
+import com.liferay.portal.test.rule.Inject;
+import com.liferay.portal.test.rule.LiferayIntegrationTestRule;
+import com.liferay.portal.test.rule.PermissionCheckerMethodTestRule;
+import com.liferay.portal.upgrade.registry.UpgradeStepRegistrator;
+import com.liferay.portal.upgrade.test.util.UpgradeTestUtil;
+
+import java.util.List;
+
+import org.junit.Before;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.runner.RunWith;
+
+/**
+ * @author Paulo Albuquerque
+ */
+@RunWith(Arquillian.class)
+public class DDMValidationUpgradeProcessDDMStructureTest
+	extends BaseCTUpgradeProcessTestCase {
+
+	@ClassRule
+	@Rule
+	public static final AggregateTestRule aggregateTestRule =
+		new AggregateTestRule(
+			new LiferayIntegrationTestRule(),
+			PermissionCheckerMethodTestRule.INSTANCE);
+
+	@Before
+	public void setUp() throws Exception {
+		_group = GroupTestUtil.addGroup();
+	}
+
+	@Override
+	protected CTModel<?> addCTModel() throws Exception {
+		_ddmStructure = DDMStructureTestUtil.addStructure(
+			_group.getGroupId(), DDMFormInstance.class.getName(),
+			DDMFormTestUtil.createDDMForm(RandomTestUtil.randomString()));
+
+		setDefinition(_ddmStructure);
+
+		return _ddmStructureLocalService.updateDDMStructure(_ddmStructure);
+	}
+
+	@Override
+	protected CTService<?> getCTService() {
+		return _ddmStructureLocalService;
+	}
+
+	@Override
+	protected void runUpgrade() throws Exception {
+		UpgradeProcess upgradeProcess = UpgradeTestUtil.getUpgradeStep(
+			_upgradeStepRegistrator, _CLASS_NAME);
+
+		upgradeProcess.upgrade();
+	}
+
+	protected void setDefinition(DDMStructure ddmStructure) {
+		DDMFormFieldValidation ddmFormFieldValidation =
+			new DDMFormFieldValidation();
+
+		ddmFormFieldValidation.setDDMFormFieldValidationExpression(
+			new DDMFormFieldValidationExpression() {
+				{
+					setName("dateRange");
+					setValue("dateRange({parameter} AND {parameter})");
+				}
+			});
+
+		DDMFormField ddmFormField = DDMFormTestUtil.createDDMFormField(
+			RandomTestUtil.randomString(), RandomTestUtil.randomString(),
+			DDMFormFieldTypeConstants.DATE, DDMFormFieldTypeConstants.DATE,
+			false, false, false);
+
+		ddmFormField.setDDMFormFieldValidation(ddmFormFieldValidation);
+
+		DDMForm ddmForm = DDMFormTestUtil.createDDMForm();
+
+		List<DDMFormField> ddmFormFields = ddmForm.getDDMFormFields();
+
+		ddmFormField.setDDMForm(ddmForm);
+
+		ddmFormFields.add(ddmFormField);
```

Comment:
@p-albuquerque why don't you just call the `ddmForm.addDDMFormField(ddmFormField);` here?